lib/faktory_worker/testing.ex

```diff
@@ -0,0 +1,288 @@
+        JobSupervisor.async_nolink(
+          Sandbox.job_supervisor(),
+          unquote(worker_mod),
+          args
+        )
+
+      timeout_duration =
+        unquote(opts)
+        |> Keyword.get_lazy(:reserve_for, &Worker.faktory_default_reserve_for/0)
+        |> Worker.reserve_timeout_duration()
+
+      receive do
+        {^job_ref, resp} ->
+          Process.demonitor(job_ref, [:flush])
+          resp
+
+        {:DOWN, ^job_ref, :process, _pid, {%{message: reason}, _stacktrace}} ->
+          {:error, reason}
+
+        {:DOWN, ^job_ref, :process, _pid, reason} ->
+          {:error, reason}
+      after
+        timeout_duration ->
+          {:error, :timeout}
+      end
+    end
+  end
+
+  @doc """
+  Resets the queue history for all job modules.
+
+  Generally, this should be called in the `setup` block for any tests that
+  use `assert_enqueued/2` or `refute_enqueued/2`, to ensure that test cases
+  don't pollute each other.
+
+  This will reset the history for _all_ job modules; if you need finer-grained
+  control over which job modules are reset (generally you shouldn't), use
+  `FaktoryWorker.Sandbox.reset/1` instead.
+  """
+  @spec reset_queues :: :ok
+  def reset_queues, do: Sandbox.reset()
+
+  @doc """
+  Assert that a job was enqueued for the given job module, optionally matching
+  a set of argument/option filters. If multiple jobs were recorded that match,
+  the assertion will still pass, so try to be as specific as possible.
+
+  ## Examples
+
+      defmodule MyApp.Test do
+        use ExUnit.Case
+
+        import FaktoryWorker.Testing
+
+        setup do
+          reset_queues()
+        end
+
+        test "enqueues" do
+          MyApp.Job.perform_async("argument")
+
+          # assert that _any_ job was enqueued for this module
+          assert_enqueued MyApp.Job
+
+          # assert that a specific job was enqueued
+          assert_enqueued MyApp.Job, args: ["argument"]
+
+          # assert on multiple arguments (order matters)
+          assert_enqueued MyApp.Job, args: ["foo", "bar"]
+
+          # assert on serializable arguments
+          assert_enqueued MyApp.Job, args: [%MyApp.User{...}]
+
+          # assert on options (only those explicitly passed to `perform_async`)
+          assert_enqueued MyApp.Job, opts: [reserve_for: 1_500]
+
+          # assert on multiple filters
+          assert_enqueued MyApp.Job, args: ["foo", "bar"], opts: [custom: %{}]
+        end
+      end
+
+  """
+  @spec assert_enqueued(module(), args: list(), opts: Keyword.t()) :: true
+  def assert_enqueued(worker_mod, filters \\ []) do
+    validate_filters!(filters)
+
+    err_message = """
+    Expected a job to be enqueued matching:
+
+    #{describe_filters(worker_mod, filters)}
+
+    but didn't find one. These jobs were enqueued for #{inspect(worker_mod)}:
+
+    #{describe_available_jobs(worker_mod)}
+    """
+
+    refute worker_mod |> Sandbox.find_jobs(filters) |> Enum.empty?(), err_message
```

Comment:
By default I think this would ideally be asserting only a single job exists that matches the `worker_mod` and `filters`. This would catch bugs where duplicate jobs are getting enqueued in the case where it isn't expected.

To support the case where duplicate jobs are expected we could maybe support passing options into the assertion that indicate how many jobs are expected. Maybe something like

```elixir
assert_enqueued(MyJob, args: [123], count: 2)
```

Any thoughts?

Reply:
I really like this idea. I'll also support passing `count: :any` to revert to the "at least one" behaviour, but leave a note in the docs that explicit counts should be preferred when possible.